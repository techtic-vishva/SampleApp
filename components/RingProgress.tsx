import { Feather } from '@expo/vector-icons'
import * as React from 'react'
import { Dimensions, Pressable } from 'react-native'
import Animated, {
	SharedValue,
	useAnimatedProps,
	useAnimatedStyle,
	withSpring,
	withTiming
} from 'react-native-reanimated'
import Svg, { Circle, Defs, Stop, LinearGradient } from 'react-native-svg'
import { chattanoogaTapWater, dichotomousHippopotamus, green } from '../constants/Colors'
import { Challenges } from '../core/services/challenges'
import useTour from '../hooks/useTour'
import { View, Text } from './Themed'

const AnimatedCircle = Animated.createAnimatedComponent(Circle)
const AnimatedSvg = Animated.createAnimatedComponent(Svg)
const minimumDisplayValue = 0.5

function hex2(c: number) {
	c = Math.round(c)
	if (c < 0) c = 0
	if (c > 255) c = 255

	var s = c.toString(16)
	if (s.length < 2) s = '0' + s

	return s
}

function color(r: number, g: number, b: number) {
	return '#' + hex2(r) + hex2(g) + hex2(b)
}

function shade(col: string, light: number) {
	light = light > 0 ? Math.min(light, 0.5) : Math.max(-0.5, light)

	// TODO: Assert that col is good and that -1 < light < 1

	var r = parseInt(col.substr(1, 2), 16)
	var g = parseInt(col.substr(3, 2), 16)
	var b = parseInt(col.substr(5, 2), 16)

	if (light < 0) {
		r = (1 + light) * r
		g = (1 + light) * g
		b = (1 + light) * b
	} else {
		r = (1 - light) * r + light * 255
		g = (1 - light) * g + light * 255
		b = (1 - light) * b + light * 255
	}

	return color(r, g, b)
}

export default ({
	progress,
	goal,
	achieved,
	sessionCount,
	onEditGoal,
	streakLength,
	sizeScalerOverride,
	challenges,
	onChallengePress
}: {
	progress: SharedValue<number>
	goal?: Duration
	achieved?: Duration
	sessionCount: number
	onEditGoal?: () => void
	streakLength?: number
	sizeScalerOverride?: number
	challenges?: Challenges[]
	onChallengePress?: () => void
}) => {
	const size = Dimensions.get('window').width * (sizeScalerOverride || 0.6)
	const r = size * 0.4
	const circumference = r * 2 * Math.PI

	const animatedStyle = useAnimatedStyle(() => {
		return {
			shadowOpacity: withTiming(progress.value >= 100 ? 0.6 : 0)
		}
	})

	const animatedProps = useAnimatedProps(() => {
		const value = progress.value || minimumDisplayValue
		const adjustedValue = value > 100 ? value % 100 : value

		return {
			strokeDashoffset: withSpring(circumference - (adjustedValue / 100) * circumference)
		}
	})

	const animatedPropsShadow = useAnimatedProps(() => {
		const value = progress.value || minimumDisplayValue
		const adjustedValue = (value > 100 ? value % 100 : value) + 0.4 // offset leading for shaddow effect

		return {
			strokeDashoffset: withSpring(circumference - (adjustedValue / 100) * circumference)
		}
	})

	const hasGoal = goal?.hours || goal?.minutes
	const hasStreak = streakLength !== undefined
	const challenge = challenges?.find((item) => item.status === 'in-progress' || item.status === 'success')

	const { TourGuideZone } = useTour('home')

	return (
		<View
			style={{
				display: 'flex',
				justifyContent: 'center',
				alignItems: 'center',
				width: '100%',
				flexDirection: 'row',
				backgroundColor: 'transparent'
			}}>
			{/* Ring center text */}
			<View
				style={{
					position: 'absolute',
					left: '35%',
					width: '30%',
					top: '33%',
					display: 'flex',
					alignItems: 'center',
					backgroundColor: 'transparent'
				}}>
				{/* Curent time */}
				<Text
					numberOfLines={1}
					adjustsFontSizeToFit={true}
					style={{
						paddingLeft: 0,
						color: 'white',
						textAlign: 'center',
						width: '100%',
						fontFamily: 'objektiv-semi-bold',
						fontSize: 40
					}}>{`${achieved?.hours ?? 0}:${(achieved?.minutes ?? 0).toString().padStart(2, '0')}`}</Text>

				{/* Session count */}
				{sessionCount > 0 && (
					<Text
						style={{
							paddingLeft: 0,
							color: 'white',
							textAlign: 'center',
							width: '100%',
							fontFamily: 'objektiv',
							fontSize: 13
						}}>
						{sessionCount} {sessionCount === 1 ? 'session' : 'sessions'}
					</Text>
				)}
			</View>

			{/* TL Complication */}
			<View
				style={{
					position: 'absolute',
					left: 0,
					width: '20%',
					display: 'flex',
					alignItems: 'flex-end',
					top: '20%',
					flexDirection: 'column',
					backgroundColor: 'transparent'
				}}>
				<Text
					style={{
						textAlign: 'center',
						color: chattanoogaTapWater,
						fontFamily: 'objektiv-semi-bold'
					}}>
					{parseInt(progress.value.toFixed(0), 10)}
					<Text style={{ fontSize: 8, color: chattanoogaTapWater, fontFamily: 'objektiv-semi-bold' }}>%</Text>
				</Text>
				<Text style={{ textAlign: 'center', color: chattanoogaTapWater, fontSize: 8 }}>Daily Time</Text>
			</View>

			{/* BL Complication */}
			{hasStreak && (
				<View
					style={{
						position: 'absolute',
						left: 0,
						width: '20%',
						display: 'flex',
						alignItems: 'flex-end',
						top: '70%',
						flexDirection: 'column',
						backgroundColor: 'transparent'
					}}>
					<View style={{ display: 'flex', flexDirection: 'row', justifyContent: 'center', alignItems: 'center' }}>
						<Text
							style={{
								textAlign: 'center',
								color: chattanoogaTapWater,
								fontFamily: 'objektiv-semi-bold'
							}}>
							{streakLength}
						</Text>
						<Text style={{ fontSize: 10, paddingLeft: 2 }}>ðŸ”¥</Text>
					</View>

					<Text style={{ textAlign: 'center', color: chattanoogaTapWater, fontSize: 8 }}>Day Streak</Text>
				</View>
			)}
			{/* activate challenge */}

			{challenge && (
				<Pressable
					onPress={onChallengePress}
					style={{
						paddingRight: 10,
						position: 'absolute',
						right: 0,
						width: '20%',
						display: 'none',
						alignItems: 'center',
						top: '70%',
						flexDirection: 'column',
						backgroundColor: 'transparent'
					}}>
					{challenge.status === 'success' && (
						<>
							<View
								style={{
									display: 'flex',
									flexDirection: 'row',
									justifyContent: 'center',
									alignItems: 'center',
									marginTop: 5
								}}>
								<Text
									numberOfLines={2}
									style={{
										fontSize: 8,
										color: chattanoogaTapWater,
										fontFamily: 'objektiv-semi-bold',
										textAlign: 'left'
									}}>
									Challenge Complete &nbsp; ðŸŽ‰
								</Text>
							</View>
						</>
					)}
					{challenge.status !== 'success' && (
						<>
							<View style={{ display: 'flex', flexDirection: 'row', justifyContent: 'center', alignItems: 'center' }}>
								<Text
									style={{
										textAlign: 'center',
										color: chattanoogaTapWater,
										fontFamily: 'objektiv-semi-bold'
									}}>
									{`${challenge?.data.progress}/${challenge?.data.max}`}
								</Text>
								<Text style={{ fontSize: 10, paddingLeft: 2 }}>{'ðŸš€'}</Text>
							</View>

							<Text
								adjustsFontSizeToFit={true}
								numberOfLines={1}
								style={{
									color: chattanoogaTapWater,
									fontSize: 8
								}}>
								Challenge Progress
							</Text>
						</>
					)}
				</Pressable>
			)}

			{/* TR Complication */}

			<Pressable
				onPress={onEditGoal}
				style={({ pressed }) => ({
					opacity: pressed ? 0.5 : 1,
					marginLeft: 'auto',
					paddingRight: 10,
					position: 'absolute',
					right: 0,
					width: '20%',
					display: 'flex',
					alignItems: 'flex-start',
					top: '20%',
					flexDirection: 'column',
					backgroundColor: 'transparent'
				})}>
				<TourGuideZone
					zone={1}
					text={`Welcome to Aro!:TITLE:And congratulations on taking this step towards developing a better relationship with your phone and in turn, a better relationship with everyone around youâ€¦including yourself.\n\nLetâ€™s show you around.`}></TourGuideZone>
				<TourGuideZone
					zone={2}
					text={
						'Daily Aro Goal:TITLE:Aro starts you with a default goal of one hour a day of phone-free time.\n\nOur algorithms will periodically evaluate your habits and recommend an increase or decrease in your goal to ensure you are motivated to fill in that ring each day.\n\nTap your goal at any time to adjust it yourself.\n'
					}>
					{hasGoal && (
						<Text adjustsFontSizeToFit={true} numberOfLines={1}>
							{typeof goal?.hours === 'number' && goal.hours > 0 && (
								<Text
									style={{
										textAlign: 'center',
										color: chattanoogaTapWater,
										fontFamily: 'objektiv-semi-bold'
									}}>
									{`${goal.hours}`}
									<Text style={{ color: chattanoogaTapWater, fontSize: 8 }}> hr </Text>
								</Text>
							)}
							<Text
								style={{
									textAlign: 'center',
									color: chattanoogaTapWater,
									fontFamily: 'objektiv-semi-bold'
								}}>
								{`${goal?.minutes ?? 0}`}
								<Text style={{ color: chattanoogaTapWater, fontSize: 8 }}> min</Text>
							</Text>
						</Text>
					)}

					<View style={{ flexDirection: 'row', alignItems: 'center' }}>
						<Text style={{ textAlign: 'center', color: chattanoogaTapWater, fontSize: 8 }}>Goal Time</Text>
						{typeof onEditGoal === 'function' && (
							<Feather name="edit-3" size={10} style={{ marginLeft: 5 }} color={chattanoogaTapWater} />
						)}
					</View>
				</TourGuideZone>
				<TourGuideZone
					zone={4}
					text={`Get Started:TITLE:Thatâ€™s all you need to get started. Simply place your phone in the smart box to start your first session. It will connect automatically. Enjoy the break from your phone.`}></TourGuideZone>
			</Pressable>
			{/* Ring */}
			<AnimatedSvg
				width={size}
				height={size}
				style={[
					{ transform: [{ rotate: '270deg' }] },
					{ shadowColor: dichotomousHippopotamus, shadowRadius: r * 0.2 },
					animatedStyle,
					{ marginLeft: 'auto', marginRight: 'auto' }
				]}>
				<Defs>
					<LinearGradient id="gradient" x1={0} x2={0} y1={0} y2={1}>
						<Stop offset="0" stopColor={dichotomousHippopotamus} stopOpacity={1} />
						<Stop offset="100%" stopColor={shade(dichotomousHippopotamus, 0.2)} stopOpacity={1} />
					</LinearGradient>
					<LinearGradient id="gradient2" x1={1} x2={0} y1={0} y2={0}>
						<Stop offset="0" stopColor={dichotomousHippopotamus} stopOpacity={1} />
						<Stop offset="100%" stopColor={shade(dichotomousHippopotamus, 0.5)} stopOpacity={1} />
					</LinearGradient>
				</Defs>

				{/* Base Ring */}
				<Circle
					stroke={'#3d2204'}
					strokeWidth={20}
					r={r}
					cx={size / 2}
					cy={size / 2}
					strokeDasharray={[circumference]}
					strokeDashoffset={0}
				/>

				{
					/* Speacial Style for > 100% */

					progress.value > 100 && (
						<>
							{/* Fixed base circle */}
							<Circle
								stroke={dichotomousHippopotamus}
								strokeWidth={Math.min(progress.value <= 100 ? 20 : 20 + 2 * Math.floor(progress.value / 100), 30)}
								r={r}
								cx={size / 2}
								cy={size / 2}
								strokeDasharray={[circumference]}
								strokeDashoffset={0}
							/>

							{/* Dropshadow Path */}
							<AnimatedCircle
								stroke={shade(dichotomousHippopotamus, -0.2)}
								strokeWidth={Math.min(progress.value <= 100 ? 20 : 20 + 2 * Math.floor(progress.value / 100), 30)}
								r={r}
								cx={size / 2}
								cy={size / 2}
								strokeDasharray={[circumference]}
								strokeDashoffset={0}
								animatedProps={animatedPropsShadow}
								strokeLinecap={'round'}
								opacity={1}
							/>
						</>
					)
				}

				<AnimatedCircle
					stroke={progress.value > 100 ? 'url(#gradient2)' : 'url(#gradient)'}
					strokeWidth={Math.min(progress.value <= 100 ? 20 : 20 + 2 * Math.floor(progress.value / 100), 30)}
					r={r}
					cx={size / 2}
					cy={size / 2}
					strokeDasharray={[circumference]}
					strokeDashoffset={0}
					animatedProps={animatedProps}
					strokeLinecap={'round'}
					opacity={1}
				/>
			</AnimatedSvg>
		</View>
	)
}
