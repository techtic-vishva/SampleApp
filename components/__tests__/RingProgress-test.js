import * as React from 'react'
import renderer from 'react-test-renderer'
import RingProgress from '../RingProgress'

it('render when passing the props', () => {
	const tree = renderer
		.create(
			<RingProgress
				progress={{ value: 0 * 100 }}
				goal={{}}
				achieved={{}}
				sessionCount={0}
				streakLength={0}
				onEditGoal={() => {
					console.log('On Edit Press')
				}}
				sizeScalerOverride={0.55}
			/>
		)
		.toJSON()
	expect(tree).toMatchSnapshot()
})
