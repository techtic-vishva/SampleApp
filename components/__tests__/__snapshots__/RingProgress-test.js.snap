// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render when passing the props 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#0b0805",
      },
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "center",
        "width": "100%",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#0b0805",
        },
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "display": "flex",
          "left": "35%",
          "position": "absolute",
          "top": "33%",
          "width": "30%",
        },
      ]
    }
  >
    <Text
      adjustsFontSizeToFit={true}
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#fff",
            "fontFamily": "objektiv",
          },
          Object {
            "color": "white",
            "fontFamily": "objektiv-semi-bold",
            "fontSize": 40,
            "paddingLeft": 0,
            "textAlign": "center",
            "width": "100%",
          },
        ]
      }
    >
      0:00
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#0b0805",
        },
        Object {
          "alignItems": "flex-end",
          "backgroundColor": "transparent",
          "display": "flex",
          "flexDirection": "column",
          "left": 0,
          "position": "absolute",
          "top": "20%",
          "width": "20%",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#fff",
            "fontFamily": "objektiv",
          },
          Object {
            "color": "#999a81",
            "fontFamily": "objektiv-semi-bold",
            "textAlign": "center",
          },
        ]
      }
    >
      0
      <Text
        style={
          Array [
            Object {
              "color": "#fff",
              "fontFamily": "objektiv",
            },
            Object {
              "color": "#999a81",
              "fontFamily": "objektiv-semi-bold",
              "fontSize": 8,
            },
          ]
        }
      >
        %
      </Text>
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#fff",
            "fontFamily": "objektiv",
          },
          Object {
            "color": "#999a81",
            "fontSize": 8,
            "textAlign": "center",
          },
        ]
      }
    >
      Daily Time
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#0b0805",
        },
        Object {
          "alignItems": "flex-end",
          "backgroundColor": "transparent",
          "display": "flex",
          "flexDirection": "column",
          "left": 0,
          "position": "absolute",
          "top": "70%",
          "width": "20%",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#0b0805",
          },
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#fff",
              "fontFamily": "objektiv",
            },
            Object {
              "color": "#999a81",
              "fontFamily": "objektiv-semi-bold",
              "textAlign": "center",
            },
          ]
        }
      >
        0
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#fff",
              "fontFamily": "objektiv",
            },
            Object {
              "fontSize": 10,
              "paddingLeft": 2,
            },
          ]
        }
      >
        ðŸ”¥
      </Text>
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#fff",
            "fontFamily": "objektiv",
          },
          Object {
            "color": "#999a81",
            "fontSize": 8,
            "textAlign": "center",
          },
        ]
      }
    >
      Day Streak
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "flex-start",
        "backgroundColor": "transparent",
        "display": "flex",
        "flexDirection": "column",
        "marginLeft": "auto",
        "opacity": 1,
        "paddingRight": 10,
        "position": "absolute",
        "right": 0,
        "top": "20%",
        "width": "20%",
      }
    }
  >
    <View
      onLayout={[Function]}
    />
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#0b0805",
            },
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#fff",
                "fontFamily": "objektiv",
              },
              Object {
                "color": "#999a81",
                "fontSize": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Goal Time
        </Text>
        <View
          color="#999a81"
          name="edit-3"
          size={10}
          style={
            Object {
              "marginLeft": 5,
            }
          }
        />
      </View>
    </View>
    <View
      onLayout={[Function]}
    />
  </View>
  <RNSVGSvgView
    bbHeight={412.50000000000006}
    bbWidth={412.50000000000006}
    focusable={false}
    height={412.50000000000006}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Array [
          Object {
            "transform": Array [
              Object {
                "rotate": "270deg",
              },
            ],
          },
          Object {
            "shadowColor": "#ed832d",
            "shadowRadius": 33.00000000000001,
          },
          Object {
            "shadowOpacity": 0,
          },
          Object {
            "marginLeft": "auto",
            "marginRight": "auto",
          },
        ],
        Object {
          "flex": 0,
          "height": 412,
          "width": 412,
        },
      ]
    }
    width={412.50000000000006}
  >
    <RNSVGGroup
      matrix={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
    >
      <RNSVGDefs>
        <RNSVGLinearGradient
          gradient={
            Array [
              0,
              -1211603,
              1,
              -943017,
            ]
          }
          gradientTransform={null}
          gradientUnits={0}
          name="gradient"
          x1={0}
          x2={0}
          y1={0}
          y2={1}
        />
        <RNSVGLinearGradient
          gradient={
            Array [
              0,
              -1211603,
              1,
              -605802,
            ]
          }
          gradientTransform={null}
          gradientUnits={0}
          name="gradient2"
          x1={1}
          x2={0}
          y1={0}
          y2={0}
        />
      </RNSVGDefs>
      <RNSVGCircle
        cx={206.25000000000003}
        cy={206.25000000000003}
        propList={
          Array [
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
          ]
        }
        r={165.00000000000003}
        stroke={4282196484}
        strokeDasharray={
          Array [
            1036.7255756846318,
            1036.7255756846318,
          ]
        }
        strokeDashoffset={null}
        strokeWidth={20}
      />
      <RNSVGCircle
        cx={206.25000000000003}
        cy={206.25000000000003}
        opacity={1}
        propList={
          Array [
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
            "strokeLinecap",
          ]
        }
        r={165.00000000000003}
        stroke={
          Array [
            1,
            "gradient",
          ]
        }
        strokeDasharray={
          Array [
            1036.7255756846318,
            1036.7255756846318,
          ]
        }
        strokeDashoffset={null}
        strokeLinecap={1}
        strokeWidth={20}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;
